# Building a Spam Filter with Naive Bayes

In this project, a spam filter will be built for SMS messages using the multinomial Naive Bayes algorithm. The goal is to write a program that classifies new messages with an accuracy greater than 80% — so one would expect that more than 80% of the new messages will be classified correctly as spam or ham (non-spam).

To train the algorithm, a dataset of 5,572 SMS messages that are already classified by humans will be used. The dataset was put together by Tiago A. Almeida and José María Gómez Hidalgo, and it can be downloaded from the [The UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection).

In [1]:
## loading libraries and data and exploring the dataset

import pandas as pd

sms_spam = pd.read_csv('SMSSpamCollection', sep='\t', header=None, names=['Label', 'SMS'])

print(sms_spam.shape)
sms_spam.head()

(5572, 2)


Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

Below, one can see that about 87% of the messages are ham, and the remaining 13% are spam. This sample looks representative, since in practice most messages that people receive are ham.

In [2]:
sms_spam['Label'].value_counts(normalize=True)

ham     0.865937
spam    0.134063
Name: Label, dtype: float64

## Training and Test Set

The data will be split into a training and a test set, where the training set accounts for 80% of the data, and the test set for the remaining 20%.

In [3]:
# Randomize the dataset
data_randomized = sms_spam.sample(frac=1, random_state=1)

# Calculate index for split
training_test_index = round(len(data_randomized) * 0.8)

# Training/Test split
training_set = data_randomized[:training_test_index].reset_index(drop=True)
test_set = data_randomized[training_test_index:].reset_index(drop=True)

print(training_set.shape)
print(test_set.shape)

(4458, 2)
(1114, 2)


Now to analyze the percentage of spam and ham messages in the training and test sets. One would expect the percentages to be close to what one has in the full dataset, where about 87% of the messages are ham, and the remaining 13% are spam.

In [4]:
print(training_set['Label'].value_counts(normalize=True))
print("\n")
print(test_set['Label'].value_counts(normalize=True))

ham     0.86541
spam    0.13459
Name: Label, dtype: float64


ham     0.868043
spam    0.131957
Name: Label, dtype: float64


## Data Cleaning

To calculate all the probabilities required by the algorithm, first a bit of data cleaning will be required to bring the data in a format that will allow one to extract easily all the information one needs.

To begin by removing all the punctuation and bringing every letter to lower case.

In [5]:
training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')
training_set['SMS'] = training_set['SMS'].str.lower()
training_set.head()

Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

## Creating the Vocabulary

creating a vocabulary, which in this context means a list with all the unique words in the training set.

In [6]:
training_set['SMS'] = training_set['SMS'].str.split()

vocabulary = []
for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))

It looks like there are 7,783 unique words in all the messages of the training set

In [7]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

In [8]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

zogtorius  anyhow  msgrcvd  personal  valentines  dokey  uneventful  \
0          0       0        0         0           0      0           0   
1          0       0        0         0           0      0           0   
2          0       0        0         0           0      0           0   
3          0       0        0         0           0      0           0   
4          0       0        0         0           0      0           0   

   radiator  total  valid  ...  08717111821  wild  wahleykkum  request  gosh  \
0         0      0      0  ...            0     0           0        0     0   
1         0      0      0  ...            0     0           0        0     0   
2         0      0      0  ...            0     0           0        0     0   
3         0      0      0  ...            0     0           0        0     0   
4         0      0      0  ...            0     0           0        0     0   

   lotr  presents  stuffs  lies  seemed  
0     0         0       0     0       0  
1     0         0       0     0       0  
2     0         0       0     0       0  
3     0         0       0     0       0  
4     0         0       0     0       0  

[5 rows x 7783 columns]

In [9]:
## combining with initial dataset
training_set_clean = pd.concat([training_set, word_counts], axis=1)
training_set_clean.head()

Label                                                SMS  zogtorius  anyhow  \
0   ham                  [yep, by, the, pretty, sculpture]          0       0   
1   ham  [yes, princess, are, you, going, to, make, me,...          0       0   
2   ham                    [welp, apparently, he, retired]          0       0   
3   ham                                           [havent]          0       0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...          0       0   

   msgrcvd  personal  valentines  dokey  uneventful  radiator  ...  \
0        0         0           0      0           0         0  ...   
1        0         0           0      0           0         0  ...   
2        0         0           0      0           0         0  ...   
3        0         0           0      0           0         0  ...   
4        0         0           0      0           0         0  ...   

   08717111821  wild  wahleykkum  request  gosh  lotr  presents  stuffs  lies  \
0            0     0           0        0     0     0         0       0     0   
1            0     0           0        0     0     0         0       0     0   
2            0     0           0        0     0     0         0       0     0   
3            0     0           0        0     0     0         0       0     0   
4            0     0           0        0     0     0         0       0     0   

   seemed  
0       0  
1       0  
2       0  
3       0  
4       0  

[5 rows x 7785 columns]

## Calculating Constants First

Now after ceaning the training set, one can begin creating the spam filter. The Naive Bayes algorithm will need to answer these two probability questions to be able to classify new messages:

$$
P(Spam | w_1,w_2, ..., w_n) \propto P(Spam) \cdot \prod_{i=1}^{n}P(w_i|Spam)
$$$$
P(Ham | w_1,w_2, ..., w_n) \propto P(Ham) \cdot \prod_{i=1}^{n}P(w_i|Ham)
$$
Also, to calculate P(wi|Spam) and P(wi|Ham) inside the formulas above, we'll need to use these equations:

$$
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
$$$$
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
$$

Some of the terms in the four equations above will have the same value for every new message. One can calculate the value of these terms once and avoid doing the computations again when a new messages comes in. Below, to use the training set to calculate:

- P(Spam) and P(Ham)
- NSpam, NHam, NVocabulary

Laplace smoothing will be set as $\alpha = 1$.

In [10]:
# Isolating spam and ham messages first
spam_messages = training_set_clean[training_set_clean['Label'] == 'spam']
ham_messages = training_set_clean[training_set_clean['Label'] == 'ham']

# P(Spam) and P(Ham)
p_spam = len(spam_messages) / len(training_set_clean)
p_ham = len(ham_messages) / len(training_set_clean)

# N_Spam
n_words_per_spam_message = spam_messages['SMS'].apply(len)
n_spam = n_words_per_spam_message.sum()

# N_Ham
n_words_per_ham_message = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham_message.sum()

# N_Vocabulary
n_vocabulary = len(vocabulary)

# Laplace smoothing
alpha = 1

## Calculating Parameters

Now that the constant terms are calculated above, one can move on with calculating the parameters $P(w_i|Spam)$ and $P(w_i|Ham)$. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

The parameters are calculated using the formulas:

$$
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
$$$$
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
$$

In [11]:
# Initiate parameters
parameters_spam = {unique_word:0 for unique_word in vocabulary}
parameters_ham = {unique_word:0 for unique_word in vocabulary}

# Calculate parameters
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()   # spam_messages already defined in a cell above
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + alpha*n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()   # ham_messages already defined in a cell above
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha*n_vocabulary)
    parameters_ham[word] = p_word_given_ham

## Classifying A New Message

Now that we have all our parameters calculated, we can start creating the spam filter. The spam filter can be understood as a function that:

- Takes in as input a new message (w1, w2, ..., wn).
- Calculates P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn).
- Compares the values of P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn), and:
  
  - If P(Ham|w1, w2, ..., wn) > P(Spam|w1, w2, ..., wn), then the message is classified as ham.
  - If P(Ham|w1, w2, ..., wn) < P(Spam|w1, w2, ..., wn), then the message is classified as spam.
  - If P(Ham|w1, w2, ..., wn) = P(Spam|w1, w2, ..., wn), then the algorithm may request human help.

In [12]:
import re

def classify(message):
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
            
    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)
    
    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal proabilities, have a human classify this!')

In [13]:
print(classify('WINNER!! This is the secret code to unlock the money: C3421.'))
print("\n")
print(classify("Sounds good, Tom, then see u there"))

P(Spam|message): 1.3481290211300841e-25
P(Ham|message): 1.9368049028589875e-27
Label: Spam
None


P(Spam|message): 2.4372375665888117e-25
P(Ham|message): 3.687530435009238e-21
Label: Ham
None


## Measuring the Spam Filter's Accuracy

The two results above look promising, but now to see how well the filter does on the test set, which has 1,114 messages.

To start by writing a function that returns classification labels instead of printing them.

In [14]:
def classify_test_set(message):    
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
    
    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

Now that there is a function that returns labels instead of printing them, one can use it to create a new column in the test set.

In [15]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

Now, to wrtie a function to measure the accuracy of the spam filter to find out how well the spam filter does.

In [16]:
correct = 0
total = test_set.shape[0]
    
for row in test_set.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct += 1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', correct/total)

Correct: 1100
Incorrect: 14
Accuracy: 0.9874326750448833


The accuracy is close to 98.74%, which is really good. This spam filter looked at 1,114 messages that it hasn't seen in training, and classified 1,100 correctly.